From 3bdb22c154f032c132bd8ab771139cfa8358b774 Mon Sep 17 00:00:00 2001
From: Avi Niddam <aniddam@habana.ai>
Date: Fri, 16 May 2025 11:32:57 +0300
Subject: [PATCH] [SW-209955] fw-loader: fail if wrong fit provided

We want to fail the fw update if a wrong FIT
(G2 instead of G3 or G3 instead of G2) is provided.
We have to check in the fw-loader Agent and in the Preboot
(to cover bad FW loader Agent case).
In this current commit, we do the changes needed for the Preboot part.
Run on G3 has been checked.

Change-Id: I63a62515e3770b0f7e1d9dcb42d8b33ad14154ab
Signed-off-by: Avi Niddam <aniddam@habana.ai>
---

diff --git a/lib/preboot/fit.c b/lib/preboot/fit.c
index cd7da3a..c9e161c 100644
--- a/lib/preboot/fit.c
+++ b/lib/preboot/fit.c
@@ -119,6 +119,8 @@
 static u32 preboot_key_idx;
 bool fit_no_ret;
 
+static const char *fit_required_device_name;
+
 void fit_clr_state(void)
 {
 	memset(images, 0, sizeof(images));
@@ -928,6 +930,41 @@
 	fit_load_print_error(err->load);
 }
 
+static int match_fit_prop_device(const void *fit, const struct fdt_property *prop, void *data)
+{
+	const char *cmpstr = (const char *)data;
+	const char *prop_name = NULL;
+
+	prop_name = fdt_get_string(fit, fdt32_ld(&prop->nameoff), NULL);
+	if (strstr(prop_name, cmpstr)) {
+		EMB_LOG_RAW(prop->data);
+		EMB_LOG_RAW("\n\r");
+		fit_required_device_name = prop->data;
+		return 0;
+	}
+
+	return -1;
+}
+
+static int fit_validate_fit_device(const void *fit)
+{
+	int rc = 0;
+
+	fit_required_device_name = NULL;
+	fit_foreach_property(fit, "/", match_fit_prop_device, HL_FIT_DEVICE_PROP);
+	if (fit_required_device_name) {
+		if (memcmp(fit_required_device_name, HL_DEVICE_TYPE, strlen(HL_DEVICE_TYPE))) {
+			EMB_LOG_ERR("Attempt to load the FIT with wrong device");
+			rc = 1;
+		} else {
+			EMB_LOG_RAW("Valid device OK");
+			EMB_LOG_RAW("\n\r");
+		}
+	}
+
+	return rc;
+}
+
 int fit_verify(void *fit, struct fit_error *err)
 {
 	int node, len;
@@ -946,6 +983,9 @@
 		return 1;
 	}
 
+	if (fit_validate_fit_device(fit))
+		return 1;
+
 	/* Check if image supports a minimum FDT version or newer */
 	if (fdt_version(fit) < FDT_MIN_SUPPORTED_VERSION) {
 		err->load = FIT_LOAD_SEC_FAILED;
